internal/common/tree.go

```diff
@@ -0,0 +1,201 @@
+package common
+
+import (
+	"slices"
+
+	"sigs.k8s.io/external-dns/endpoint"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+)
+
+type DNSTreeNode struct {
+	Name     string
+	Children []*DNSTreeNode
+	DataSets []DNSTreeNodeData
+}
+
+type DNSTreeNodeData struct {
+	RecordType       string
+	SetIdentifier    string
+	RecordTTL        endpoint.TTL
+	Labels           endpoint.Labels
+	ProviderSpecific endpoint.ProviderSpecific
+	Targets          []string
+}
+
+// RemoveNode removes a node from a tree.
+// If the node was the only child of the parent node,
+// the parent will be removed as well unless the parent is a root node
+func (n *DNSTreeNode) RemoveNode(deleteNode *DNSTreeNode) {
+	// store indexes of dead branches
+	var deadBranches []int
+	var deadDataSets []string
+
+	if deleteNode == nil {
+		return
+	}
+
+	for i, node := range n.Children {
+		if node.Name == deleteNode.Name {
+			n.Children = append(n.Children[:i], n.Children[i+1:]...)
+
+			// we removed child, but need to clean up data sets
+			for j, dataset := range n.DataSets {
+				index := slices.Index(dataset.Targets, deleteNode.Name)
+				if index > 0 {
```

Comment:
what if the index is 0? should this be >=0 ? looks like `index` will return -1 if not present